### PR TITLE
Relax PasswordGenerator API and add retry mech.

### LINKED ISSUE #152: Password generator does not always generate valid password.
@serac  @dfish3r 

I have a password validator which also includes a rule like IllegalSequenceRule. I have also a generator but PasswordGenerator is taking into account only CharacterRule-s.

So what might happen is that a password is generated with exactly same character rules as my validator is using, but it might happen that it will create a password which also includes illegal sequence. So what I generate with the character rules is just not valid when I test it with my validator which uses same character rules (plus IllegalSequenceRule).

This is quite hard to detect in practice but a simple test which generates e.g. 1 million passwords yields some passwords (like ... couple hundreds, it really depends) which will not pass the validator.

In my software, I can not simply allow this to happen. What I am doing is that when I generate a password, I will run it via a validator and if it passes I return it to a user, but when it fails, I just try to create a new one. I am basically running it in a loop until a password is generated in such a way that it will pass the validator.

Now, in great majority of cases, a valid password will be generated just on the first attempt but this is just annoying stuff to deal with. 

MV:@@kjihG2]MZ8w{zs{

but you see that Ghijk (reversed kjihG) is illegal because it is a sequence.

### PR DESCRIPTION
**Pull request by serac**
Relax generate API to accept a list of Rule so that rules that are naturally part of password validation can also be specified for password generation. The common use case is IllegalSequenceRule. Check the generated password against the given ruleset and try again up to 2 times to generate a valid password.

Resolves #152

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
   <groupId>org.passay</groupId>
   <artifactId>passay</artifactId>
   <packaging>jar</packaging>
-  <version>1.6.5-SNAPSHOT</version>
+  <version>1.7.0-SNAPSHOT</version>
```
Comment (dfish3r) [inline pom.xml:10]: I believe this version bump is unnecessary. The API change you made does not break binary API compatibility.

File: src/main/java/org/passay/PasswordGenerator.java [hunk 2]
```diff
@@ -17,10 +17,15 @@
  */
 public class PasswordGenerator
 {
+  /** Retry password generation at most this many times. */
+  private static final int RETRY_LIMIT = 2;
```
Comment (dfish3r) [inline src/main/java/org/passay/PasswordGenerator.java:21]: I think this will need to be configurable. It's a fine default, but ultimately something that deployers will want to tweak.

File: src/main/java/org/passay/PasswordGenerator.java [hunk 3]
```diff
@@ -124,4 +155,10 @@ protected void randomize(final CharBuffer buffer)
       buffer.put(i, c);
     }
   }
+
+
+  private void validate(final String password, final List<Rule> rules)
```
Comment (dfish3r) [inline src/main/java/org/passay/PasswordGenerator.java:160]: Why is this here?

File: pom.xml (hunk 1 above)
Comment (serac) [reply pom.xml:10]: The following code will not compile in this branch:

    final List<CharacterRule> ruleList = Arrays.asList(
            new CharacterRule(EnglishCharacterData.LowerCase),
            new CharacterRule(EnglishCharacterData.UpperCase));
    final PasswordGenerator gen = new PasswordGenerator();
    gen.generatePassword(10, ruleList)
    
Given that this would presumably be a common way to invoke the password generator, then it's not API compatible and the version bump is warranted. I will admit that I was surprised that `japicmp-maven-plugin` didn't complain, but I did not investigate.

File: src/main/java/org/passay/PasswordGenerator.java (hunk 3 above)
Comment (serac) [reply src/main/java/org/passay/PasswordGenerator.java]: Vestigial code in flight that I forgot to remove. Pushed a fix.

File: pom.xml (hunk 1 above)
Comment (dfish3r) [reply pom.xml:10]: One day I'll remember that generics are invariant. Created a PR to preserve API compatibility:
https://github.com/vt-middleware/passay/pull/155

(japicmp runs in the verify phase, so you might not have exercised it.)

File: src/main/java/org/passay/PasswordGenerator.java (hunk 2 above)
Comment (dfish3r) [reply src/main/java/org/passay/PasswordGenerator.java:21]: Fair enough, I'm fine with a wait and see approach here.

File: pom.xml (hunk 1 above)
Comment (serac) [reply pom.xml]: Accepted, thanks!

File: src/main/java/org/passay/PasswordGenerator.java (hunk 2 above)
Comment (smiklosovic) [reply src/main/java/org/passay/PasswordGenerator.java:21]: @serac I would like to have it configurable too. It might be a range from 1 to ... 10? Something similar.  I would rather

Comment (smiklosovic) [reply src/main/java/org/passay/PasswordGenerator.java:21]: This is what I did in Apache Cassandra:

https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/guardrails/CassandraPasswordGenerator.java#L57-L77

The default is 100 attempts :D Yeah I know, this is an absolute overkill and some figure like "10" would be probably enough but I do not think that having the number of attempts so big is problematic because it will generate a valid password in 99% in two attempts anyway. That configuration property is not something a users are going to use frequently if ever.

At the same time, you do not have any real influence over the probability that the password will not be invalid. You may have a super bad luck and it would need to generate it three times, who knows? If you generate passwords to be 30 characters long, the probability of generating illegal sequence just raises. I would rather generate it by more attempts rather then failed completely.
